xds/src/main/java/io/grpc/xds/FaultFilter.java

```diff
@@ -215,23 +218,48 @@ public ClientInterceptor buildClientInterceptor(
           // TODO(https://github.com/grpc/grpc-java/issues/7868)
           callExecutor = MoreExecutors.directExecutor();
         }
-        if (finalDelayNanos != null && finalAbortStatus != null) {
-          return new DelayInjectedCall<>(
-              finalDelayNanos, callExecutor, scheduler, callOptions.getDeadline(),
-              Suppliers.ofInstance(
-                  new FailingClientCall<ReqT, RespT>(finalAbortStatus, callExecutor)));
-        }
-        if (finalAbortStatus != null) {
-          return new FailingClientCall<>(finalAbortStatus, callExecutor);
+        if (finalDelayNanos != null) {
+          Supplier<? extends ClientCall<ReqT, RespT>> callSupplier;
+          if (finalAbortStatus != null) {
+            callSupplier = Suppliers.ofInstance(
+                new FailingClientCall<ReqT, RespT>(finalAbortStatus, callExecutor));
+          } else {
+            callSupplier = new Supplier<ClientCall<ReqT, RespT>>() {
+              @Override
+              public ClientCall<ReqT, RespT> get() {
+                return next.newCall(method, callOptions);
+              }
+            };
+          }
+          final DelayInjectedCall<ReqT, RespT> delayInjectedCall = new DelayInjectedCall<>(
+              finalDelayNanos, callExecutor, scheduler, callOptions.getDeadline(), callSupplier);
+
+          final class DeadlineInsightForwardingCall extends ForwardingClientCall<ReqT, RespT> {
+            @Override
+            protected ClientCall<ReqT, RespT> delegate() {
+              return delayInjectedCall;
+            }
+
+            @Override
+            public void start(Listener<RespT> listener, Metadata headers) {
+              Listener<RespT> finalListener =
+                  new SimpleForwardingClientCallListener<RespT>(listener) {
+                    @Override
+                    public void onClose(Status status, Metadata trailers) {
+                      if (status.getCode().equals(Code.DEADLINE_EXCEEDED)) {
+                        status = status.augmentDescription(
+                            String.format("Additional delay injected: %d ns", finalDelayNanos));
```

Comment:
Why not just add it to `DelayInjectedCall` instead of wrapping another layer of the call? `Status.augmentDescription()` isn't a good thing as we discussed before. Also, the more wrapping, the harder to read and debug the code. In this case, it can be easily avoided.

Reply:
> Why not just add it to DelayInjectedCall instead of wrapping another layer of the call?

Because DelayInjectedCall can not extend `ForwordingCall` and `DelayedClientCall` as the same time, so wrapping is inevitable. 

>Status.augmentDescription() isn't a good thing as we discussed before.

I was keeping in mind of that when working on the PR, but there is no obvious better and simple alternative.

Reply:
Fine, not too bad.